gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BomConflictCheckTask.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.plugins;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import netflix.nebula.dependency.recommender.provider.RecommendationProviderContainer;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.TaskAction;
+
+
+public class BomConflictCheckTask extends DefaultTask {
+
+    private final File propsFile;
+
+    @Inject
+    public BomConflictCheckTask(File propsFile) {
+        this.propsFile = propsFile;
+    }
+
+    @Input
+    public final Map<String, String> getMavenBomRecommendations() {
+        return getProject().getExtensions()
+                .getByType(RecommendationProviderContainer.class)
+                .getMavenBomProvider()
+                .getRecommendations();
+    }
+
+    @InputFile
+    public final File getPropsFile() {
+        return propsFile;
+    }
+
+    @TaskAction
+    public final void checkBomConflict() {
+        Map<String, String> recommendations = getMavenBomRecommendations();
+        List<Conflict> conflicts = new LinkedList<>();
+        Set<String> artifacts = BaselineVersions.getResolvedArtifacts(getProject());
+        Map<String, String> resolvedConflicts = new HashMap<>();
+        BaselineVersions.checkVersionsProp(getPropsFile(),
+                (propName, propVersion) -> {
+                    String regex = propName.replaceAll("\\*", ".*");
+                    artifacts.stream()
+                            .filter(artifactName -> artifactName.matches(regex))
+                            .forEach(artifactName -> resolvedConflicts.put(artifactName, propName));
+
+                    recommendations.entrySet()
+                            .stream()
+                            .filter(entry -> entry.getKey().matches(regex))
+                            .forEach(entry -> {
+                                conflicts.add(new Conflict(propName, propVersion, entry.getKey(), entry.getValue()));
+                                resolvedConflicts.remove(entry.getKey());
```

Comment:
you already have an artifact here

Reply:
I see what you're trying to do here, but I find the add, then remove logic confusing.

The yes/no conditions could be collapsed into multiple .filter() invocations and I think that reads better:

```java
Map<String, String> resolvedConflicts = BaselineVersions.readVersionsProps(getPropsFile())
                .stream()
                .flatMap(pair -> {
                    String propName = pair.getLeft();
                    String propVersion = pair.getRight();
                    String regex = propName.replaceAll("\\*", ".*");

                    Set<String> recommendationConflicts = recommendations
                            .entrySet()
                            .stream()
                            .filter(entry -> entry.getKey().matches(regex))
                            .map(entry -> {
                                conflicts.add(new Conflict(propName, propVersion, entry.getKey(), entry.getValue()));
                                return entry.getKey();
                            })
                            .collect(Collectors.toSet());

                    return artifacts.stream()
                            .filter(artifactName -> artifactName.matches(regex))
                            .filter(artifactName -> !recommendationConflicts.contains(artifactName))
                            .map(artifactName -> Pair.of(artifactName, propName));
                })
                .collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
```

Reply:
Adressed the review. Kept the pair. What is the data structure that you suggest instead @robert3005 ?